CMakeLists.txt

```diff
@@ -207,19 +207,34 @@ endif()
 
 # If available and wanted, use compiler cache.
 if(LIBCELLML_COMPILER_CACHE)
-  if(WIN32)
-    if(CLCACHE_EXE)
-      set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
-      execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
-                      RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
-      if(RESULT EQUAL 0)
-        set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
-      endif()
+  if(BUILDCACHE_EXE)
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${BUILDCACHE_EXE})
+  elseif(CLCACHE_EXE)
+    set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
+    execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
+                    RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
+    if(RESULT EQUAL 0)
+      set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
     endif()
   else()
-    if(CCACHE_EXE)
-      set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
-    endif()
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+  endif()
+endif()
+
+# On Windows, if we are using MSVC then replace some compiler flags.
+if(MSVC)
+  # If we are treating warnings as errors then remove the /W3 option since it
+  # will eventually be replaced with the /W4 and /WX options (see
+  # target_warnings_as_errors()).
+  if(LIBCELLML_TREAT_WARNINGS_AS_ERRORS)
+    replace_compiler_flag("/W3" "")
+  endif()
+
+  # Replace the /Zi option (i.e. enable debugging information), if present, with
+  # the /Z7 option (i.e. enable old-style debugging information) otherwise
+  # caching (be it with buildcache or clcache) won't work.
+  if(BUILDCACHE_EXE OR CLCACHE_EXE)
+    replace_compiler_flag("/Zi" "/Z7")
   endif()
 endif()
```

Comment:
This should all go into a macro called something like `apply_compiler_cache_settings` and 
 put in `cmake/macros.cmake` .

Reply:
Removing the `/W3` compiler flag has nothing to do with it, but ok let's make things more difficult to maintain. Done in commit 23ce6714.

Reply:
I think this makes it infinitely easier to maintain we have decluttered the main entry point and have a descriptive label for what this part contains.